source/common/common/trie.h

```diff
@@ -0,0 +1,118 @@
+  TrieNode() : TrieNode(nullptr) {}
+  ~TrieNode() {}
+
+  /**
+   * Insert a value into the trie.
+   * @param key the name associated with the value.
+   * @param value the value to store on the node associated with name
+   */
+  void emplace(const Key& key, const Value& value);
+
+  /**
+   * Lookup a value by key.
+   * @param key the path name to look up
+   * @return a std::pair consisting of the value if found and a boolean
+   *         indicating if the match was exact.
+   */
+  std::pair<Value*, bool> find(const Key& key) const;
+
+private:
+  std::pair<Value*, bool> find(std::vector<Key>& path_components) const;
+  void emplace(std::vector<Key>& path_components, const Value& value);
+  void set_value(const Value& value) { value_ = value; }
+
+  // By holding pointers to the TriedNodes we don't have to worry that they're
+  // incomplete types.
+  std::unordered_map<Key, std::unique_ptr<TrieNode<Key, Value, Tokenizer>>> children_;
+  // value_ is being held as an opaque value and mutating it doesn't mutate the
+  // container.
+  mutable Value value_;
+  Tokenizer tokenizer_;
+};
+
+template <class Key, class Value, class Tokenizer>
+void TrieNode<Key, Value, Tokenizer>::emplace(const Key& key, const Value& value) {
+  std::vector<Key> path_components = tokenizer_.tokenize(key);
+  reverseComponentVector(path_components);
+  emplace(path_components, value);
+}
+
+// TODO(tschroed): Make this iterative rather than recursive.
+template <class Key, class Value, class Tokenizer>
+void TrieNode<Key, Value, Tokenizer>::emplace(std::vector<Key>& path_components,
+                                              const Value& value) {
+  if (path_components.empty()) {
+    return;
+  }
+  Key name(path_components.back());
+  path_components.pop_back();
+  TrieNode<Key, Value, Tokenizer>* node = nullptr;
+  auto it = children_.find(name);
+  if (children_.find(name) == children_.end()) {
+    node = new TrieNode<Key, Value, Tokenizer>(nullptr);
+    children_.emplace(name, std::move(std::unique_ptr<TrieNode<Key, Value, Tokenizer>>(node)));
+  } else {
+    node = it->second.get();
+  }
+  if (path_components.empty()) {
+    node->set_value(value);
+  } else {
+    node->emplace(path_components, value);
+  }
+}
+
+template <class Key, class Value, class Tokenizer>
+std::pair<Value*, bool> TrieNode<Key, Value, Tokenizer>::find(const Key& key) const {
+  std::vector<Key> path_components = tokenizer_.tokenize(key);
+  reverseComponentVector(path_components);
+  return find(path_components);
+}
+
+// TODO(tschroed): Make this iterative rather than recursive.
+template <class Key, class Value, class Tokenizer>
+std::pair<Value*, bool>
+TrieNode<Key, Value, Tokenizer>::find(std::vector<Key>& path_components) const {
+  if (path_components.empty()) {
+    // An exact match requires that we have a non-false value at the very end of
+    // our path traversal.
+    return std::make_pair(&value_, value_ ? true : false);
```

Comment:
FYI this is from another comment of mine. There is no real requirement that Value be a pointer, in fact, it would be better if it didn't have to be. The only real requirement is that it is constructable from nullptr. I'm kind of fried right now, but optimally would do away with the nullptr requirement and just use Optional<> or a sentinel some type, then I think this make sense. Or have the template actually require shared_ptr<Value> if want it to only ever be a pointer. Would potentially like to consider some other solutions here but don't have brain power to offer anything right now.